### ISSUE TITLE
htmlToBlocks collapses whitespace inside pre tags

### ISSUE THREAD

**Issue by mjm** [sanity-io/sanity#1814]
**Describe the bug**

Before parsing HTML, the `htmlToBlocks` function in `@sanity/block-tools` deletes all extra whitespace in the string of HTML. It seems to rely on this in general to not produce extra empty paragraph blocks, but this makes it impossible to use this to import multiline code blocks correctly.

**To Reproduce**

Here's an isolated script that I can run with `sanity exec` that reproduces the behavior:

```js
import createSchema from 'part:@sanity/base/schema-creator'
import schemaTypes from 'all:part:@sanity/base/schema-type'
import { htmlToBlocks } from '@sanity/block-tools'
import { JSDOM } from 'jsdom'

const schema = createSchema({
  name: 'blog',
  types: [
    ...schemaTypes,
    {
      name: 'bodyPortableText',
      type: 'array',
      of: [{ type: 'block' }]
    }
  ]
})

const bodyContentType = schema.get('bodyPortableText')

const html = `
<p>Foo bar</p>
  <pre><code>const a = 1;
const b = 2;
const c = 3;</code></pre>
`

console.log(
  JSON.stringify(
    htmlToBlocks(html, bodyContentType, {
      parseHtml: html => new JSDOM(html).window.document
    }),
    null,
    4
  )
)
```

which produces:

```json
[
    {
        "_type": "block",
        "markDefs": [],
        "style": "normal",
        "children": [
            {
                "_type": "span",
                "marks": [],
                "text": "Foo bar",
                "_key": "c57eb14820040"
            }
        ],
        "_key": "c57eb1482004"
    },
    {
        "_type": "block",
        "markDefs": [],
        "style": "normal",
        "children": [
            {
                "_type": "span",
                "marks": [
                    "code"
                ],
                "text": "const a = 1; const b = 2; const c = 3;",
                "_key": "8240b633021b0"
            }
        ],
        "_key": "8240b633021b"
    }
]
```

**Expected behavior**

The second block with the code should have the spacing/newlines that the original HTML included, as they are significant to how the HTML is rendered.

In my actual use, I'm trying to create a custom rule to transform pre tags into a custom block type, but the nodes that I get have already lost the whitespace, so there's not really a good way to get them back.

**Which versions of Sanity are you using?**

```
❯ sanity versions
@sanity/cli             1.149.4 (latest: 1.149.6)
@sanity/base            1.149.5 (up to date)
@sanity/cli             1.149.4 (latest: 1.149.6)
@sanity/code-input      1.149.0 (up to date)
@sanity/components      1.149.5 (latest: 1.149.6)
@sanity/core            1.149.4 (latest: 1.149.6)
@sanity/dashboard       1.149.5 (up to date)
@sanity/default-layout  1.149.5 (up to date)
@sanity/default-login   1.149.0 (up to date)
@sanity/desk-tool       1.149.5 (up to date)
@sanity/studio-hints    1.149.5 (up to date)
```

**What operating system are you using?**

macOS 10.15.3

**Which versions of Node.js / npm are you running?**

```
❯ npm -v && node -v
6.13.6
v12.11.1
```

**Additional context**

I haven't found a good workaround for this yet. It's possible to skip the pre-processing that removes this whitespace by passing the original HTML directly to JSDOM in `parseHtml` instead of the `html` provided to that function, but if I do that, I end up with many empty paragraph blocks in the output.

**Comment by artemagvanian** [comment]
@mjm have you managed to discover any workarounds yet? I have already spent quite a bit of time trying to fight this behavior while migrating a huge dataset, and it looks like I am a bit stuck

**Comment by mjm** [comment]
@artemagvanian I must have managed to hack something together for this, because I'm pretty sure I was using this to migrate my existing website into Sanity, and I did succeed in doing that. That said, it's been over a year, so I don't really remember what I did.

Here's the script I used: https://github.com/mjm/mattmoriarity-com/blob/master/studio/bin/transform-blog-posts.js but looking at it, it's not obvious what I did to get around this problem. It's possible I directly edited something in node_modules to make it work. Sorry I can't be more help than that.

**Comment by artemagvanian** [comment]
@mjm thanks a lot for sharing! I managed to overcome the issue by using the normalizeBlock function with a partial block since I had just raw text data to migrate into blocks.

**Comment by straube** [comment]
I haven't seen this issue before and ended up opening another one: https://github.com/sanity-io/sanity/issues/3977. It's likely a duplicate.

**Comment by sjelfull** [comment]
Thanks for reporting this issue. This should be fixed in #4540 and was released in [v3.12.0](https://github.com/sanity-io/sanity/releases/tag/v3.12.0).
